kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableClusteringE2ESite.scala

```diff
@@ -0,0 +1,172 @@
+      // Verify the clustering domain metadata is written
+      verifyClusteringDomainMetadata(snapshot)
+    }
+  }
+
+  test("update a table with clustering columns should be blocked") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      createEmptyTable(engine, tablePath, testPartitionSchema)
+      val table = Table.forPath(engine, tablePath)
+      val txnBuilder = table.createTransactionBuilder(engine, testEngineInfo, WRITE)
+
+      val ex = intercept[TableAlreadyExistsException] {
+        txnBuilder
+          .withClusteringColumns(engine, testClusteringColumns.asJava)
+          .build(engine)
+      }
+      assert(
+        ex.getMessage
+          .contains("Table already exists, but provided new clustering columns"))
+    }
+  }
+
+  test("insert into clustered table - table create from scratch") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val testData = Seq(Map.empty[String, Literal] -> dataClusteringBatches1)
+
+      val commitResult = appendData(
+        engine,
+        tablePath,
+        isNewTable = true,
+        testPartitionSchema,
+        clusteringCols = testClusteringColumns,
+        data = testData)
+
+      verifyCommitResult(commitResult, expVersion = 0, expIsReadyForCheckpoint = false)
+      verifyCommitInfo(tablePath, version = 0, operation = WRITE)
+      verifyWrittenContent(
+        tablePath,
+        testPartitionSchema,
+        dataClusteringBatches1.flatMap(_.toTestRows))
+
+      val table = Table.forPath(engine, tablePath)
+      verifyClusteringDomainMetadata(table.getLatestSnapshot(engine).asInstanceOf[SnapshotImpl])
+    }
+  }
+
+  test("insert into clustered table - already existing table") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val testData = Seq(Map.empty[String, Literal] -> dataClusteringBatches1)
+
+      {
+        val commitResult0 = appendData(
+          engine,
+          tablePath,
+          isNewTable = true,
+          testPartitionSchema,
+          clusteringCols = testClusteringColumns,
+          data = Seq(Map.empty[String, Literal] -> dataClusteringBatches1))
+
+        val expData = dataClusteringBatches1.flatMap(_.toTestRows)
+
+        verifyCommitResult(commitResult0, expVersion = 0, expIsReadyForCheckpoint = false)
+        verifyCommitInfo(tablePath, version = 0, operation = WRITE)
+        verifyWrittenContent(tablePath, testPartitionSchema, expData)
+      }
+      {
+        val commitResult1 = appendData(
+          engine,
+          tablePath,
+          data = Seq(Map.empty[String, Literal] -> dataClusteringBatches2))
+
+        val expData = dataClusteringBatches1.flatMap(_.toTestRows) ++
+          dataClusteringBatches2.flatMap(_.toTestRows)
+
+        verifyCommitResult(commitResult1, expVersion = 1, expIsReadyForCheckpoint = false)
+        verifyCommitInfo(tablePath, version = 1, partitionCols = null, operation = WRITE)
+        verifyWrittenContent(tablePath, testPartitionSchema, expData)
+      }
+    }
+  }
```

Comment:
negative cases where data files with no stats for the clustering columns are added?

Reply:
case where invalid clustering column is provided?